services/server/server_windows.go

```diff
@@ -20,12 +20,71 @@ package server
 
 import (
 	"context"
+	"unsafe"
 
+	"github.com/containerd/containerd/log"
 	srvconfig "github.com/containerd/containerd/services/server/config"
 	"github.com/containerd/ttrpc"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
 )
 
-func apply(_ context.Context, _ *srvconfig.Config) error {
+func getPriorityClass(priorityClassName string) uint32 {
+	var priorityClassMap = map[string]uint32{
+		"IDLE_PRIORITY_CLASS":         uint32(windows.IDLE_PRIORITY_CLASS),
+		"BELOW_NORMAL_PRIORITY_CLASS": uint32(windows.BELOW_NORMAL_PRIORITY_CLASS),
+		"NORMAL_PRIORITY_CLASS":       uint32(windows.NORMAL_PRIORITY_CLASS),
+		"ABOVE_NORMAL_PRIORITY_CLASS": uint32(windows.ABOVE_NORMAL_PRIORITY_CLASS),
+		"HIGH_PRIORITY_CLASS":         uint32(windows.HIGH_PRIORITY_CLASS),
+		"REALTIME_PRIORITY_CLASS":     uint32(windows.REALTIME_PRIORITY_CLASS),
+	}
+	return priorityClassMap[priorityClassName]
+}
+
+// addCurrentProcessToJobObjectAndSetPriorityClass creates a new Job Object
+// (https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects),
+// adds the current process to the job object, and specifies the priority
+// class for the job object to the specified value.
+// A job object is used here so that any spawned processes such as CNI or
+// shim binaries are created at the specified thread priority class.
+// Running containerd and shim binaries with above normal / high priority
+// can help improve responsiveness on machines with high CPU utilization.
+func addCurrentProcessToJobObjectAndSetPriorityClass(pc uint32) error {
+	job, err := windows.CreateJobObject(nil, nil)
+	if err != nil {
+		return err
+	}
+	limitInfo := windows.JOBOBJECT_BASIC_LIMIT_INFORMATION{
+		LimitFlags:    windows.JOB_OBJECT_LIMIT_PRIORITY_CLASS,
+		PriorityClass: pc}
+	if _, err := windows.SetInformationJobObject(
+		job,
+		windows.JobObjectBasicLimitInformation,
+		uintptr(unsafe.Pointer(&limitInfo)),
+		uint32(unsafe.Sizeof(limitInfo))); err != nil {
+		return err
+	}
+	if err := windows.AssignProcessToJobObject(job, windows.CurrentProcess()); err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
Can just return `windows.AssignProcessToJobObject(job, windows.CurrentProcess())`, although I'd prefer if this error was wrapped instead as win32 errors are never helpful haha

Reply:
I'll wrap it

Reply:
updated